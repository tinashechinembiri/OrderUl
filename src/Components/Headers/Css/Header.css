.headers{
    display:grid; 
    grid-template-areas: "logo nav";
    border: 2px solid rgba(255,255,255,0.1);
    background-color: rgb(248, 247, 248); 
    
    
}
.nav-items{
   grid-area: nav;
   display: grid; 
   grid-template-columns: repeat(20, auto); 
   align-items: left;
   justify-items: left;
   
}
.Logo{
    width: 80px;
    height: 80px;
    align-items: left
    ;
}
.nav-items_div{ 
    display: flex;
    justify-content:space-evenly;
}
.nav-items_div li  { 
    list-style-type: none;
    justify-content: space-between;
}