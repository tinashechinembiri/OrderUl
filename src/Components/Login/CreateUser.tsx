import React, {  Fragment } from "react";
//https://kentcdodds.com/blog/using-fetch-with-type-script
const createuser = () => { 

}