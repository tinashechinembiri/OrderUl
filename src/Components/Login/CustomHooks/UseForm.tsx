import  { useState } from 'react'; 
import {omit} from 'lodash'; 
 interface Loginvalues {
    username?:string | undefined, 
    password?:string | undefined
}
type IunionType = Loginvalues; 
const useForm = () => {
    const [values, setValues] = useState <IunionType>(); 
    const [errors, setErrors] = useState<IunionType>(); 

    const handleChange = (event:React.ChangeEvent<HTMLInputElement>) => {
        event.persist(); 
        let name:string = event.target.name; 
        let val:string = event.target.value; 
        validate( name, val); 
        // validate values in state
     
       // setInputs(inputs => ({...inputs, [event.target.name]: event.target.value}));
        setValues({
            ...values ,  
            [name]:val,
        }); 

        

    }
  
    
    const validate =( name :string, value:string ) => {
        
        switch(name)
        {
            //https://stackoverflow.com/questions/12018245/regular-expression-to-validate-username
            case'username': 
            let usernameerror = new RegExp( /^\w+@[a-zA-Z_]+?\.[a-zA-Z]{2,3}$/); 
            if (!usernameerror.test(value))
            {
                console.log(usernameerror)
                setErrors({
                    ...errors, 
                    username:'Username atleast have 5 letters'
                }); 
                console.log(errors)
            }
            else{
                let newObj = omit(errors, "username");
                setErrors(newObj);
            }
          
            break; 
            case'password': 
            if (!new RegExp(/^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9]).{8,}$/).test(value))
            {
                setErrors({
                    ...errors, 
                    password:'Password should contains atleast 8 charaters and containing uppercase,lowercase and numbers'
                }); 
            }
            else{
                let newObj = omit(errors, "password");
                setErrors(newObj);
            }
            break; 
            case'email':
            break; 
            

            default:
                break; 
        }
    }
    return{
        handleChange,
        values, 
        errors     
    }
}
export default useForm;     
