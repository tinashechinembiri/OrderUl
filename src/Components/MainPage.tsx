
import WhatsOn from "./HomePage/WhatsOn";

const MainPage = () => 
{
 
        return(
            <div className="mainpage">
            <WhatsOn/>
            </div>   
        );   
    
}
export default MainPage; 