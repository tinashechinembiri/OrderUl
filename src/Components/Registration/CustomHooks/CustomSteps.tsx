import React, {useState} from 'react';

 const CustomSteps = () => {
    const [currentSteps, SetcurrentSteps] = useState(1); 

    const _next = () => {
        let currentstep = currentSteps<3 ?currentSteps +1 : currentSteps;
           console.log(currentstep)    
         SetcurrentSteps(currentstep); 
    }
    const _prev =() => {
        let currentstep = currentSteps>1 ? currentSteps -1: currentSteps; 
        console.log(currentstep)
        SetcurrentSteps(currentstep); 
    }

    return{
        currentSteps, 
        _next,
        _prev
    }

}
export default CustomSteps; 