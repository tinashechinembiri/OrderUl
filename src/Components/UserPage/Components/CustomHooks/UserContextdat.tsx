
import React, {createContext,ReactNode, useEffect, useContext, useReducer} from "react";
import  {UserInterface} from'../Interface/UserInterface'; 
import {getuserdata,updateProfile} from '../../service/Userprofileservice';


const _userprofileData = {
    data: {
        user_id: "",
        email: "",
        password: "",
        accounts: {
            Address: {
                addressname: "",
                postcode: "",
                county: "",
                country: ""
            },  
            accountid: "",
            firstname: "",
            lastname: "",
            phone: undefined,
        }
    },
     _errors:undefined


}
interface EditData {
    email : string, 
    [key: string|number] : string |number|undefined, 

}

type userprofileData = {
    data : UserInterface, 
    _errors? : object|null
}

const Usercontext = createContext<userprofileData> (_userprofileData); 
const UserprofileDispatcher = createContext <React.Dispatch<actionType>>(() =>null); 
export type actionType = {type:'userprofile',  userprofile:userprofileData, }| {type:'edit', userprofile:any }; 
// type AppContextValue = {
//     userProfile: UserInterface;
//     // type, you get when hovering over `setState` from `useState`
//     SetuserProfile: React.Dispatch<React.SetStateAction<UserInterface>>
//   };
const UserProfileReducer = ( _userprofileData : any, action :actionType) => {

    switch(action.type)
    {
           case  'userprofile':
               const {userprofile} = action 
               if (userprofile._errors)
               {
                   console.log('errors')
                   return {
                       ..._userprofileData, 
                    
                       _errors:{..._userprofileData._errors, _errors:userprofile._errors}
                   }
               }
              return {
                  data : userprofile.data,
                  _errors:{..._userprofileData._errors, _errors:null}
              }
            case 'edit': 
              if (userprofile._errors)
              {
                 return {
                     ..._userprofileData, 
                 
                     _errors:{..._userprofileData._errors, _errors:userprofile._errors}
                 }

              }
                return {
                     data : userprofile.data,
                     _errors:{..._userprofileData._errors, _errors:null}
                     }

              default : 
              return _userprofileData; 
              
    }
}

export function UseuserProfileState () {
    const context = useContext(Usercontext)
    if (context === undefined)
    {
        throw new Error('hasnt been created'); 
    }
    return context; 

}
export function  UserprofileAuthDispatch () {
    const context = useContext(UserprofileDispatcher); 
    if (context === undefined)
    {
        throw new Error('hasnt been created'); 
    }
    return context; 
}
const userprofileAysncHandler = (dispatch:React.Dispatch<actionType>) => {
    return async (action :actionType) => {
        let payload; 
        switch (action.type)
        {
            case 'edit':
                const  updateprofile =  await updateProfile(action.userprofile);
                payload = {...updateProfile  }
                //dispatch

                
            break; 
            case'userprofile':

                 payload = {...action.userprofile}; 
             
                 dispatch ({type:'userprofile', userprofile:payload})
            break


        }
    }
}

export const  Usercontextprovider =({children}:{ children: ReactNode}) =>{
    const [value, dispatch] = useReducer (UserProfileReducer, _userprofileData )
    
  
    
    return(
        <Usercontext.Provider value={value}>
        <UserprofileDispatcher.Provider value={userprofileAysncHandler(dispatch)}>
             {children}
        </UserprofileDispatcher.Provider>
        </Usercontext.Provider>
    )
}


export {Usercontext}
export default Usercontextprovider; 