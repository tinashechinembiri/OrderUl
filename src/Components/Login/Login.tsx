import React, {  Fragment,  } from "react";
import {useNavigate } from "react-router-dom";
import LoginService from "./Service/LoginService";
import useForm from './CustomHooks/UseForm'; 
import './Css/Login.css'; 
import {useAuthDispatch, useAuthstate} from'../../Helpers/Context'; 
export interface Loginvalues {
    username:string , 
    password:string
}

interface IloginService{
  error?: string  
}


const Login = () => {
    const dispatch = useAuthDispatch(); 
    const  { user,loading, _errors} = useAuthstate(); 
    const navigate = useNavigate();
    
   // const [username, setUsername] = useState(""); 
    //const [password, setPassword] = useState('')
// /https://www.bezkoder.com/react-hooks-jwt-auth/
// css : https://codepen.io/FlorinPop17/pen/vPKWjd
      //https://www.sitepoint.com/how-to-migrate-a-react-app-to-typescript/  

    const  handleSubmit = async ( e:React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();        //const values; 
       if (Object.keys(errors as object).length === 0 && Object.keys(values as object).length!== 0)
        {
           const _loginservice =  await LoginService.Login(values as {username:string, password:string},dispatch ) as IloginService; 
           if (!_loginservice?.error)
           {
               
            return navigate('/homepage')

           }
           return; 

        }
    }
   
    const {handleChange , values, errors}  = useForm(); 
 
    return(
        <Fragment> 
            <div className="form-body">
                  {
                        (_errors|| null) && <h3>{(_errors|| null)}</h3>
                    }
            <form className="form-class" onSubmit={handleSubmit}>
                <label>username:</label>
                  
                <input
                    className="input-username"
                    name="username"
                    type='email'
                    placeholder="enter your username"
                    onChange={handleChange}
                />
                    {
                     ( errors?.username|| null) && <h3> {(errors?.username || null)}</h3>
                    }
                <div>
                    <label>Password: </label>
                    <input
                    className="input-password"
                     type='password'
                     placeholder="enter a password"
                     name="password"
                     onChange={handleChange}
                    />
                    {
                      errors?.password && <h3> {errors?.password}</h3>
                    }
                </div>
                <button className="btn" type="submit"> Login</button>
            </form>
            </div>
        </Fragment>
    )

}
export default Login; 