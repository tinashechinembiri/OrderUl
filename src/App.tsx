
import React ,{ useState, useEffect } from 'react';
import {
  BrowserRouter as Router,
  Routes,
  Route
} from "react-router-dom";

import MainPage from './Components/MainPage'; 
import Login from'./Components/Login/Login'; 
import './App.css';
import LoginService from './Components/Login/Service/LoginService';
import Headers from './Components/Headers/HeadersNav';
import {AuthPrrovider} from'./Helpers/Context'
import {useAuthDispatch, useAuthstate} from'./Helpers/Context'
import Usercontextprovider from './Components/UserPage/Components/CustomHooks/UserContextdat'
import CreateUser from './Components/Registration/CreateUser'
import UserPage from './Components/UserPage/UserPage'
import EditPage  from './Components/UserPage/EditPage'
function App() {
  //const [currentUser , setCurrentuser] = useState <any|null>(); 
  const dispatch = useAuthDispatch(); 
  const logOut = () => {
    LoginService.logout(dispatch); 
  }
  
  return (
    <div className="App">
      <AuthPrrovider>
      <div> 
      <Headers
        logouts={logOut}
      />
      </div>
      <Usercontextprovider>
        <Router>
        <Routes>
          <Route path='/' element={<MainPage/>}/>
          <Route path='/login'  element={<Login/>}/>
          <Route path='/register' element={<CreateUser/>}/>
          <Route path='/userprofile' element={<UserPage/>}/>
          <Route path='/edit/:editpage' element={<EditPage/>}/>      
        </Routes>
        </Router>
      </Usercontextprovider>
      </AuthPrrovider>
    </div>
  );
}

export default App;
