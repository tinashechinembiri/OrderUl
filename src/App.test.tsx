import React, {ReactNode}from 'react';
import { render, screen,  } from '@testing-library/react';
//import MainPage from './Components/MainPage'; 
import Login from './Components/Login/Login'; 
import {Usercontext} from './Components/UserPage/Components/CustomHooks/UserContextdat'
import Header from  './Components/Headers/HeadersNav'
import '@testing-library/jest-dom/extend-expect';

import App from './App';
import {MemoryRouter } from 'react-router-dom'

import {create} from 'react-test-renderer';
import {  createMemoryHistory } from 'history';
//https://kentcdodds.com/blog/common-mistakes-with-react-testing-library

import MainPage from './Components/MainPage';
jest.mock('./Components/MainPage.tsx'); 

jest.mock('./Components/Headers/HeadersNav.tsx')
// jest.mock('./Components/Login/Login.tsx')
// jest.mock('./Components/Registration/CreateUser.tsx')
// jest.mock('./Components/UserPage/UserPage.tsx')
// jest.mock('./Components/UserPage/EditPage.tsx')

describe ('app ts', ()=>{
    let mockMainpagecomponent:jest.MockedFunction < typeof MainPage>
    let mockHeadercomponent:jest.MockedFunction < typeof Header>
beforeEach(() => {
    mockMainpagecomponent = MainPage as jest.MockedFunction < typeof MainPage>;
    mockHeadercomponent = Header as jest.MockedFunction < typeof Header>;
    mockMainpagecomponent.mockImplementation ( () => {
        return (
            <>
            HomePageMock
            </>
        )
    }); 
    mockHeadercomponent.mockImplementation(() => { return (<div> PageHeaderMock</div>)})
}); 
afterEach (()=> {
    jest.clearAllMocks(); 
})
it('create a snapshot of the app', () => {

    const component = create (<App/>)
   let tree =  component.toJSON(); 
   expect (tree).toMatchSnapshot(); 
});
it('should render the main page on the router', ()=> {  
    render(<App/> ); 
expect(screen.getByText("PageHeaderMock")).toBeInTheDocument();
expect(screen.getByText("HomePageMock")).toBeInTheDocument();
});
it('check the current router', ()=> {  
 
});
})