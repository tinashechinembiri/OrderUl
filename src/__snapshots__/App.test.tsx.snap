// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app ts create a snapshot of the app 1`] = `
<div
  className="App"
>
  <div>
    <div>
       PageHeaderMock
    </div>
  </div>
  HomePageMock
</div>
`;
